In [4]:
#importing libraries

import os
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from sklearn.preprocessing import OneHotEncoder

In [5]:
# setting directory
os.chdir(r'C:\buan6383')

In [6]:
# reading data

data = pd.read_csv(r'data\transactions.csv')

In [7]:
# Use pd.get_dummies to do onehotencoding and make separate columns

data = pd.get_dummies(data, columns=['Product'])
data.head()

Transaction  Product_Bow  Product_Candy Bar  Product_Deodorant  \
0        12359        False               True              False   
1        12362        False              False              False   
2        12362        False              False              False   
3        12365        False              False              False   
4        12371         True              False              False   

   Product_Greeting Cards  Product_Magazine  Product_Markers  \
0                   False             False            False   
1                   False             False            False   
2                   False             False            False   
3                   False             False            False   
4                   False             False            False   

   Product_Pain Reliever  Product_Pencils  Product_Pens  Product_Perfume  \
0                  False            False         False            False   
1                   True            False         False            False   
2                   True            False         False            False   
3                  False            False         False            False   
4                  False            False         False            False   

   Product_Photo Processing  Product_Prescription Med  Product_Shampoo  \
0                     False                     False            False   
1                     False                     False            False   
2                     False                     False            False   
3                     False                     False            False   
4                     False                     False            False   

   Product_Soap  Product_Toothbrush  Product_Toothpaste  \
0         False               False               False   
1         False               False               False   
2         False               False               False   
3         False               False                True   
4         False               False               False   

   Product_Wrapping Paper  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False

In [8]:
data.shape

(459258, 18)

In [9]:
# This tab takes a while to run

data = data.groupby('Transaction').agg(any)
data.shape

(200000, 17)

In [10]:
#Renaming columns to remove Product_ from column names

data = data.rename(columns=lambda x: x[len('Product_'):] if x.startswith('Product_') else x)

In [11]:
data.head()

Bow  Candy Bar  Deodorant  Greeting Cards  Magazine  Markers  \
Transaction                                                                   
12359        False       True      False           False     False    False   
12362        False      False      False           False     False    False   
12365        False      False      False           False     False    False   
12371         True      False      False           False     False    False   
12380        False      False      False            True     False    False   

             Pain Reliever  Pencils   Pens  Perfume  Photo Processing  \
Transaction                                                             
12359                False    False  False    False             False   
12362                 True    False  False    False             False   
12365                False    False  False    False             False   
12371                False    False  False    False             False   
12380                False    False  False    False             False   

             Prescription Med  Shampoo   Soap  Toothbrush  Toothpaste  \
Transaction                                                             
12359                   False    False  False       False       False   
12362                   False    False  False       False       False   
12365                   False    False  False       False        True   
12371                   False    False  False       False       False   
12380                   False    False  False       False       False   

             Wrapping Paper  
Transaction                  
12359                 False  
12362                 False  
12365                 False  
12371                 False  
12380                 False

In [12]:
# Writing data to a csv file
data.to_csv('group05transactions01.csv', index=True)

In [13]:
data.shape

(200000, 17)

In [14]:
# Copying data to trax
trax = data.copy()

In [15]:
trax.head(9)

Bow  Candy Bar  Deodorant  Greeting Cards  Magazine  Markers  \
Transaction                                                                   
12359        False       True      False           False     False    False   
12362        False      False      False           False     False    False   
12365        False      False      False           False     False    False   
12371         True      False      False           False     False    False   
12380        False      False      False            True     False    False   
12383        False      False      False           False     False    False   
12386        False      False      False           False     False    False   
12392        False      False      False           False      True    False   
12401        False       True      False           False      True    False   

             Pain Reliever  Pencils   Pens  Perfume  Photo Processing  \
Transaction                                                             
12359                False    False  False    False             False   
12362                 True    False  False    False             False   
12365                False    False  False    False             False   
12371                False    False  False    False             False   
12380                False    False  False    False             False   
12383                 True    False  False    False             False   
12386                 True    False  False    False             False   
12392                False    False  False    False             False   
12401                False     True  False    False             False   

             Prescription Med  Shampoo   Soap  Toothbrush  Toothpaste  \
Transaction                                                             
12359                   False    False  False       False       False   
12362                   False    False  False       False       False   
12365                   False    False  False       False        True   
12371                   False    False  False       False       False   
12380                   False    False  False       False       False   
12383                   False    False  False       False       False   
12386                   False    False  False       False       False   
12392                   False     True  False       False       False   
12401                   False    False  False       False       False   

             Wrapping Paper  
Transaction                  
12359                 False  
12362                 False  
12365                 False  
12371                 False  
12380                 False  
12383                 False  
12386                 False  
12392                 False  
12401                 False

In [16]:
# Using the apriori algorithm to find the frequent itemsets with 
# Minimum support threshold of 1%

fi = apriori(trax, min_support = 0.01, use_colnames = True)
fi

support                                 itemsets
0   0.054645                                    (Bow)
1   0.171005                              (Candy Bar)
2   0.146885                         (Greeting Cards)
3   0.241305                               (Magazine)
4   0.026700                          (Pain Reliever)
5   0.134925                                (Pencils)
6   0.143575                                   (Pens)
7   0.089960                                (Perfume)
8   0.058480                       (Photo Processing)
9   0.014505                       (Prescription Med)
10  0.033800                                (Shampoo)
11  0.043025                                   (Soap)
12  0.067350                             (Toothbrush)
13  0.160425                             (Toothpaste)
14  0.050990                         (Wrapping Paper)
15  0.011340                        (Bow, Toothbrush)
16  0.043660              (Greeting Cards, Candy Bar)
17  0.040535                    (Candy Bar, Magazine)
18  0.033015                     (Candy Bar, Pencils)
19  0.012265                        (Candy Bar, Pens)
20  0.039780                  (Toothpaste, Candy Bar)
21  0.036335               (Greeting Cards, Magazine)
22  0.029240                (Greeting Cards, Pencils)
23  0.032080             (Greeting Cards, Toothpaste)
24  0.031630                      (Pencils, Magazine)
25  0.021150                         (Magazine, Pens)
26  0.017190                      (Perfume, Magazine)
27  0.016455             (Photo Processing, Magazine)
28  0.013175                   (Toothbrush, Magazine)
29  0.031665                   (Toothpaste, Magazine)
30  0.024560                    (Toothpaste, Pencils)
31  0.010315                       (Toothpaste, Pens)
32  0.021820                    (Toothbrush, Perfume)
33  0.016665    (Greeting Cards, Candy Bar, Magazine)
34  0.013175  (Greeting Cards, Toothpaste, Candy Bar)
35  0.012005           (Candy Bar, Pencils, Magazine)
36  0.013720        (Toothpaste, Candy Bar, Magazine)
37  0.011390         (Toothpaste, Candy Bar, Pencils)
38  0.010955      (Greeting Cards, Pencils, Magazine)
39  0.011945   (Greeting Cards, Toothpaste, Magazine)

### 40 items are frequent and out of these 15 are singletons

### We have 15 singletons which means 2 columns are in less than 1% of the transactions

### 1% is around 2000 transactions

In [17]:
# identifying association rules with minimum threshold of 10%

rules = association_rules(fi, metric="confidence", min_threshold=0.1)
rules

antecedents            consequents  antecedent support  \
0                          (Bow)           (Toothbrush)            0.054645   
1                   (Toothbrush)                  (Bow)            0.067350   
2               (Greeting Cards)            (Candy Bar)            0.146885   
3                    (Candy Bar)       (Greeting Cards)            0.171005   
4                    (Candy Bar)             (Magazine)            0.171005   
5                     (Magazine)            (Candy Bar)            0.241305   
6                    (Candy Bar)              (Pencils)            0.171005   
7                      (Pencils)            (Candy Bar)            0.134925   
8                   (Toothpaste)            (Candy Bar)            0.160425   
9                    (Candy Bar)           (Toothpaste)            0.171005   
10              (Greeting Cards)             (Magazine)            0.146885   
11                    (Magazine)       (Greeting Cards)            0.241305   
12              (Greeting Cards)              (Pencils)            0.146885   
13                     (Pencils)       (Greeting Cards)            0.134925   
14              (Greeting Cards)           (Toothpaste)            0.146885   
15                  (Toothpaste)       (Greeting Cards)            0.160425   
16                     (Pencils)             (Magazine)            0.134925   
17                    (Magazine)              (Pencils)            0.241305   
18                        (Pens)             (Magazine)            0.143575   
19                     (Perfume)             (Magazine)            0.089960   
20            (Photo Processing)             (Magazine)            0.058480   
21                  (Toothbrush)             (Magazine)            0.067350   
22                  (Toothpaste)             (Magazine)            0.160425   
23                    (Magazine)           (Toothpaste)            0.241305   
24                  (Toothpaste)              (Pencils)            0.160425   
25                     (Pencils)           (Toothpaste)            0.134925   
26                  (Toothbrush)              (Perfume)            0.067350   
27                     (Perfume)           (Toothbrush)            0.089960   
28   (Greeting Cards, Candy Bar)             (Magazine)            0.043660   
29    (Greeting Cards, Magazine)            (Candy Bar)            0.036335   
30         (Candy Bar, Magazine)       (Greeting Cards)            0.040535   
31              (Greeting Cards)  (Candy Bar, Magazine)            0.146885   
32  (Greeting Cards, Toothpaste)            (Candy Bar)            0.032080   
33   (Greeting Cards, Candy Bar)           (Toothpaste)            0.043660   
34       (Toothpaste, Candy Bar)       (Greeting Cards)            0.039780   
35          (Candy Bar, Pencils)             (Magazine)            0.033015   
36         (Candy Bar, Magazine)              (Pencils)            0.040535   
37           (Pencils, Magazine)            (Candy Bar)            0.031630   
38       (Toothpaste, Candy Bar)             (Magazine)            0.039780   
39        (Toothpaste, Magazine)            (Candy Bar)            0.031665   
40         (Candy Bar, Magazine)           (Toothpaste)            0.040535   
41       (Toothpaste, Candy Bar)              (Pencils)            0.039780   
42         (Toothpaste, Pencils)            (Candy Bar)            0.024560   
43          (Candy Bar, Pencils)           (Toothpaste)            0.033015   
44     (Greeting Cards, Pencils)             (Magazine)            0.029240   
45    (Greeting Cards, Magazine)              (Pencils)            0.036335   
46           (Pencils, Magazine)       (Greeting Cards)            0.031630   
47  (Greeting Cards, Toothpaste)             (Magazine)            0.032080   
48    (Greeting Cards, Magazine)           (Toothpaste)            0.036335   
49        (Toothpaste, Magazine)       (Greeting Cards)            0.03

### We find 50 rules with a minimum confidence of 10% 


In [18]:
# Sorting first by lift and then by confidence

# Highest lift is 3.60


rules2 = rules.sort_values(['lift', 'confidence'], ascending =[False, False])
rules2

antecedents            consequents  antecedent support  \
27                     (Perfume)           (Toothbrush)            0.089960   
26                  (Toothbrush)              (Perfume)            0.067350   
0                          (Bow)           (Toothbrush)            0.054645   
1                   (Toothbrush)                  (Bow)            0.067350   
30         (Candy Bar, Magazine)       (Greeting Cards)            0.040535   
31              (Greeting Cards)  (Candy Bar, Magazine)            0.146885   
42         (Toothpaste, Pencils)            (Candy Bar)            0.024560   
29    (Greeting Cards, Magazine)            (Candy Bar)            0.036335   
49        (Toothpaste, Magazine)       (Greeting Cards)            0.031665   
39        (Toothpaste, Magazine)            (Candy Bar)            0.031665   
32  (Greeting Cards, Toothpaste)            (Candy Bar)            0.032080   
46           (Pencils, Magazine)       (Greeting Cards)            0.031630   
34       (Toothpaste, Candy Bar)       (Greeting Cards)            0.039780   
45    (Greeting Cards, Magazine)              (Pencils)            0.036335   
37           (Pencils, Magazine)            (Candy Bar)            0.031630   
36         (Candy Bar, Magazine)              (Pencils)            0.040535   
43          (Candy Bar, Pencils)           (Toothpaste)            0.033015   
41       (Toothpaste, Candy Bar)              (Pencils)            0.039780   
40         (Candy Bar, Magazine)           (Toothpaste)            0.040535   
48    (Greeting Cards, Magazine)           (Toothpaste)            0.036335   
33   (Greeting Cards, Candy Bar)           (Toothpaste)            0.043660   
2               (Greeting Cards)            (Candy Bar)            0.146885   
3                    (Candy Bar)       (Greeting Cards)            0.171005   
28   (Greeting Cards, Candy Bar)             (Magazine)            0.043660   
44     (Greeting Cards, Pencils)             (Magazine)            0.029240   
47  (Greeting Cards, Toothpaste)             (Magazine)            0.032080   
35          (Candy Bar, Pencils)             (Magazine)            0.033015   
13                     (Pencils)       (Greeting Cards)            0.134925   
12              (Greeting Cards)              (Pencils)            0.146885   
9                    (Candy Bar)           (Toothpaste)            0.171005   
8                   (Toothpaste)            (Candy Bar)            0.160425   
7                      (Pencils)            (Candy Bar)            0.134925   
6                    (Candy Bar)              (Pencils)            0.171005   
38       (Toothpaste, Candy Bar)             (Magazine)            0.039780   
14              (Greeting Cards)           (Toothpaste)            0.146885   
15                  (Toothpaste)       (Greeting Cards)            0.160425   
20            (Photo Processing)             (Magazine)            0.058480   
25                     (Pencils)           (Toothpaste)            0.134925   
24                  (Toothpaste)              (Pencils)            0.160425   
10              (Greeting Cards)             (Magazine)            0.146885   
11                    (Magazine)       (Greeting Cards)            0.241305   
4                    (Candy Bar)             (Magazine)            0.171005   
5                     (Magazine)            (Candy Bar)            0.241305   
16                     (Pencils)             (Magazine)            0.134925   
17                    (Magazine)              (Pencils)            0.241305   
22                  (Toothpaste)             (Magazine)            0.160425   
23                    (Magazine)           (Toothpaste)            0.241305   
21                  (Toothbrush)             (Magazine)            0.067350   
19                     (Perfume)             (Magazine)            0.089960   
18                        (Pens)             (Magazine)            0.14



The rule {Perfume} --> {Toothbrush} has the highest lift of 3.601

Since the lift is symmetric {Toothbrush} --> {Perfume} has the same lift



A : Antecedent, C : Consequent

Lift is calculated using the following formula: 

Support (A & C) / Support (A)*Support (C)


Another way to calculate it is: Confidence (Rule) / Support (C)



For {Perfume} --> {Toothbrush} we calculate lift this way: 

0.021820 / (0.089960 * 0.067350) = 3.601370

## Leverage

Leverage is another measure of independence, it is a measure of deviation from independence. 

It is calculated using the following formula:
    
Leverage = Support(A & C) - (Support(A) * Support(C))

A leverage of 0 imples A and C are Statistically independent

Greater than 0 implies Positive Correlation

Less than 0 implies Negative Correlation


Calculation: 
    
    0.021820 - (0.067350 * 0.089960) = 0.015761
    
    
This suggests a slight positive relationship between perfume and toothbrush purchases.

## Conviction

Conviction is another measure of dependency, it checks the dependence of consequent on the antecedent. 

A conviction of 1 implies independence. 

Formula: 
    
    1 - Support(C) / (1 - Confidence(A --> C)) = Conviction
    
Calculation: 
    
    The conviction for rule 27 :
    (1- .067350)/(1-.242552)=1.231306


    The conviction for rule 36:
    (1- .08996)/(1-.32398)=1.346172


In [19]:
# Getting top 5 rules with highest confidence
rules.nlargest(n=5,columns=['confidence'])

antecedents       consequents  antecedent support  \
42         (Toothpaste, Pencils)       (Candy Bar)            0.024560   
29    (Greeting Cards, Magazine)       (Candy Bar)            0.036335   
39        (Toothpaste, Magazine)       (Candy Bar)            0.031665   
30         (Candy Bar, Magazine)  (Greeting Cards)            0.040535   
32  (Greeting Cards, Toothpaste)       (Candy Bar)            0.032080   

    consequent support   support  confidence      lift  leverage  conviction  \
42            0.171005  0.011390    0.463762  2.711980  0.007190    1.545947   
29            0.171005  0.016665    0.458649  2.682078  0.010452    1.531344   
39            0.171005  0.013720    0.433286  2.533762  0.008305    1.462810   
30            0.146885  0.016665    0.411126  2.798966  0.010711    1.448723   
32            0.171005  0.013175    0.410692  2.401637  0.007689    1.406726   

    zhangs_metric  
42       0.647160  
29       0.650802  
39       0.625124  
30       0.669879  
32       0.602960

Some insights we can gather from these results are:

- The same 4 items appear several times in the top 5 rules, sometimes simply switching up what is considered an antecedent versus a consequent.
- 5/5 of these rules include a Candy Bar. Candy Bars appear in 17% of the transactions. 
- Magazines, Greeting Cards and Toothpaste each appear in 3/5 rules. 
- The most frequently bought items in the store are Magazines (24%), Candy Bars (17%), Toothpaste (16%), and Greeting Cards (15%), in that order. 
- While their frequency could be alarming, we can notice that the confidence of each rule (range: 0.41-0.46) is usually higher than the support of the consequents. So, these rules are likely helping us make correct associations and increase the chance of purchasing the consequents to ~40%. 
- Lift is >1 in all rules – hence, we can assume they all have a positive correlation. 
- Conviction ranges from 1.4-1.55. This tells us that, when we assume the antecedents and consequents of the rules above are related, if these two events are actually independent, we’d be wrong about 1.4-1.55 times more often. 


In [20]:
## Getting top 5 rules with highest lift

rules.nlargest(n=5,columns=['lift'])

antecedents       consequents  antecedent support  \
27              (Perfume)      (Toothbrush)            0.089960   
26           (Toothbrush)         (Perfume)            0.067350   
0                   (Bow)      (Toothbrush)            0.054645   
1            (Toothbrush)             (Bow)            0.067350   
30  (Candy Bar, Magazine)  (Greeting Cards)            0.040535   

    consequent support   support  confidence      lift  leverage  conviction  \
27            0.067350  0.021820    0.242552  3.601370  0.015761    1.231306   
26            0.089960  0.021820    0.323979  3.601370  0.015761    1.346172   
0             0.067350  0.011340    0.207521  3.081236  0.007660    1.176877   
1             0.054645  0.011340    0.168374  3.081236  0.007660    1.136755   
30            0.146885  0.016665    0.411126  2.798966  0.010711    1.448723   

    zhangs_metric  
27       0.793732  
26       0.774490  
0        0.714499  
1        0.724232  
30       0.669879

Some insights we can gather from these results are:
- Candy Bars appear in 3/5 rules as a consequent. 
- Toothbrushes appear in 2/5 rules. 
- Magazines appear in 2/5 rules. 
- Candy Bars and Magazines items are some of the most frequently sold items at the store. 
- Lift is >1 in all rules – hence, we can assume they all have a positive correlation.  
- Confidence is higher than the consequent support – this means that anytime the antecedents are purchased, the chance of the consequent getting purchased increases. 
- This is the second time that  {pencils, toothpaste} --> {candy bar}, makes an appearance. 
- Rule where {greeting cards, magazines}-->{candy bar}, also made an appearance in the last sort

In [21]:
# Getting top 5 rules with highest leverage

rules_sortc = rules.sort_values(['leverage'], ascending =[False])
rules_sortc.head(10)

antecedents            consequents  antecedent support  \
2             (Greeting Cards)            (Candy Bar)            0.146885   
3                  (Candy Bar)       (Greeting Cards)            0.171005   
26                (Toothbrush)              (Perfume)            0.067350   
27                   (Perfume)           (Toothbrush)            0.089960   
8                 (Toothpaste)            (Candy Bar)            0.160425   
9                  (Candy Bar)           (Toothpaste)            0.171005   
30       (Candy Bar, Magazine)       (Greeting Cards)            0.040535   
31            (Greeting Cards)  (Candy Bar, Magazine)            0.146885   
29  (Greeting Cards, Magazine)            (Candy Bar)            0.036335   
6                  (Candy Bar)              (Pencils)            0.171005   

    consequent support   support  confidence      lift  leverage  conviction  \
2             0.171005  0.043660    0.297239  1.738191  0.018542    1.179626   
3             0.146885  0.043660    0.255314  1.738191  0.018542    1.145604   
26            0.089960  0.021820    0.323979  3.601370  0.015761    1.346172   
27            0.067350  0.021820    0.242552  3.601370  0.015761    1.231306   
8             0.171005  0.039780    0.247966  1.450053  0.012347    1.102338   
9             0.160425  0.039780    0.232625  1.450053  0.012347    1.094087   
30            0.146885  0.016665    0.411126  2.798966  0.010711    1.448723   
31            0.040535  0.016665    0.113456  2.798966  0.010711    1.082253   
29            0.171005  0.016665    0.458649  2.682078  0.010452    1.531344   
6             0.134925  0.033015    0.193065  1.430903  0.009942    1.072050   

    zhangs_metric  
2        0.497810  
3        0.512294  
26       0.774490  
27       0.793732  
8        0.369675  
9        0.374393  
30       0.669879  
31       0.753386  
29       0.650802  
6        0.363260

In [22]:
# Dropping redundant rows
rules_sortc=rules_sortc.drop([3,27,9,31])
rules_sortc.head(5)

antecedents       consequents  antecedent support  \
2             (Greeting Cards)       (Candy Bar)            0.146885   
26                (Toothbrush)         (Perfume)            0.067350   
8                 (Toothpaste)       (Candy Bar)            0.160425   
30       (Candy Bar, Magazine)  (Greeting Cards)            0.040535   
29  (Greeting Cards, Magazine)       (Candy Bar)            0.036335   

    consequent support   support  confidence      lift  leverage  conviction  \
2             0.171005  0.043660    0.297239  1.738191  0.018542    1.179626   
26            0.089960  0.021820    0.323979  3.601370  0.015761    1.346172   
8             0.171005  0.039780    0.247966  1.450053  0.012347    1.102338   
30            0.146885  0.016665    0.411126  2.798966  0.010711    1.448723   
29            0.171005  0.016665    0.458649  2.682078  0.010452    1.531344   

    zhangs_metric  
2        0.497810  
26       0.774490  
8        0.369675  
30       0.669879  
29       0.650802

- Some insights we can gather from these results are:
- Candy Bar appears in 4/5 rules. 
- Greeting Cards appear in 3/5 rules.
- It is now the third time that Rule #29 makes an appearance.
- Rule #26 has a high lift result, and is showing up for the second time in our analysis. 
- Rule #26 {Perfume} --> {Toothbrush} is an interesting rule because it includes items we have not encountered in our top association rules yet. 
- Both of these items are less frequently purchased than the other products that consistently show up in our analysis. For instance, toothbrushes are bought about 7% of the time. But when someone has bought perfume, the chance of a toothbrush purchase more than triples, to 24%. 
- While we do not have access to the prices of these items, we must recognize that perfumes are often pricier than toothbrushes (definitely pricier than the top 4 products, like candy bars or magazines). Hence, this rule is extra appealing due to its sale of a higher-end product.
- Through a more practical lens, we noticed that toothpaste and perfume fall within the realm of “self-care” and “hygiene.” A person who takes care of their dental hygiene enough to get a new toothbrush is likely to also care about other aspects of their hygiene, like their perceived smell. Hence, a rational hypothesis is that the customers who buy perfume and toothbrushes could also be interested. 

In [23]:
# Getting top 5 rules with highest conviction

rules.nlargest(n=5,columns=['conviction'])

antecedents       consequents  antecedent support  \
42         (Toothpaste, Pencils)       (Candy Bar)            0.024560   
29    (Greeting Cards, Magazine)       (Candy Bar)            0.036335   
39        (Toothpaste, Magazine)       (Candy Bar)            0.031665   
30         (Candy Bar, Magazine)  (Greeting Cards)            0.040535   
32  (Greeting Cards, Toothpaste)       (Candy Bar)            0.032080   

    consequent support   support  confidence      lift  leverage  conviction  \
42            0.171005  0.011390    0.463762  2.711980  0.007190    1.545947   
29            0.171005  0.016665    0.458649  2.682078  0.010452    1.531344   
39            0.171005  0.013720    0.433286  2.533762  0.008305    1.462810   
30            0.146885  0.016665    0.411126  2.798966  0.010711    1.448723   
32            0.171005  0.013175    0.410692  2.401637  0.007689    1.406726   

    zhangs_metric  
42       0.647160  
29       0.650802  
39       0.625124  
30       0.669879  
32       0.602960

Some insights we can gather from these results are:
- We have seen every single one of these rules before in our analysis. In fact, these rules include only the top 5 most frequently sold products in the store. 
- Most of these rules have Candy Bar as their consequent.
- Confidence is higher than the consequent support for every single one of these rules. Hence, the consequents have a better chance of being purchased when they are bundled with the antecedents than when they are not. 

Overall, we reviewed a little less than 20 rules. Most of these 20 rules included the same 4 products (Magazine, Candy Bar, Toothpaste, Greeting Cards) which are the ones that are most frequently purchased. Hence, it feels redundant to utilize 10+ rules trying to predict the sale of a Candy Bar or a Magazine when it is already likely that they will purchase it. 